File: OCAT - Search - PA.py
import sys, os, shutil
#	loads the sys, os, and shutil modules for Python to use

droppedPath	=	sys.argv[1].rsplit("\\", 1)[0] + "\\"

scriptPath	=	sys.argv[0].rsplit("\\", 1)[0] + "\\"
#	sys.argv is a list of the arguments passed to the script
#		the first item in the list is the path to the script itself
#		the second and subsequent items are the paths of files dropped onto the script
#	rsplit is a command to split a string at the designated pattern, optionally into a specified number of pieces
#		it is used here to remove the file names, getting just their paths (and the slashes are added back for droppedPath, as they are needed later)

# switches to determine if this calls for a single-GPU or multi-GPU configuration
if "Review" in droppedPath.rsplit("OCAT Data")[0].rsplit("\\", 2)[1]:
	TYPE	=	"SINGLE"
#	when working with review data, it will always be a single-GPU situation
elif "OCAT Data" in droppedPath.rsplit("\\", 4)[2:4]:
	TYPE	=	"MULTI"
#	if the dropped file is in the OCAT Data folder, it is the multi-GPU situation of collecting together all data for a specific quality level
#	if the dropped file is in a folder one below OCAT Data, then this may be a multi-API situation and so I want the quality level fixed
else:
	TYPE	=	"SINGLE"
#	all other non-review situations are single-GPU so just using else
#		there is technically the situation of collecting all data for a single-GPU and single-API (if applicable) but this is one I rarely use so it is not configured here
#		plus it would require checking the presence of multiple APIs or not

mQUA	=	"High"
#	the quality to filter by for multi-GPU situations

RelPath	=	droppedPath.split("OCAT Data")[0] + "OCAT Data\\"
#	splits the path at the OCAT Data folder, grabs the first portion, then adds the folder name back to the path

listfile	=	[]
#	creates an empty list for appending things too

for paths, folders, files in os.walk(droppedPath):
#	os.walk is a generator function, which means it does not itself produce an output but works with the for loop
#	os.walk will generate lists for three variables; the paths, folders, and file names
	for file in files:
#		for loop that goes through each file in the list of files
		if file.startswith("OCAT-"):
#			the OCAT CSVs all start with OCAT-, so this makes sure it is only these files that are used
			listfile.append((str(paths).replace(RelPath, "") + "\\" + str(file)).replace("\\\\", "\\"))
#				adds on to the listfile list path strings for each CSV file
#					removes the RelPath from the path strings, so the path is just from the GPU folder down
#	produces a list of all OCAT CSVs with the directory information

listsplit	=	[file.split("\\") for file in listfile]
#	creates a multi-dimensional list by spliting the path lists according to the folder names

listmap	=	[]
#	creates an empty list that will be used to map the information to the appropriate places
for line in listsplit:
#	loop to go through each line in the listsplit list
	listmapL	=	[line[0], "", "", ""]
#		listmap Line list variable created with four places
#			order is GPU, API, Quality, File name
#			first element of the listsplit lines is always the GPU
	for i in range(1, len(line)):
#		loop going through the places in the line
		listmapL[len(listmapL) - i]	=	line[len(line) - i]
#			this applies the value from the line to the listmapL variable
#			it goes backwards, and the reason is so that if the API value is not present, it will be left as an empty string
#			a key thing is that it uses the length of the line from listsplit, because when there is no API folder, the length is shorter, so the API place in listmap is untouched
#				it is also important that Python starts its indexing at 0, so by using a range starting at 1, it will not include the first element here
	listmap.append(listmapL)
#		adds the listmapL line to the listmap variable
	
#	this will map the values I need to the appropriate locations in a list
#		[GPU, API, Quality, File]
#	if no API change is made, the element will be blank


GPUs, QUAs	=	[], []
#	create empty lists for storing the lists of relevant values

for item in listmap:
	GPUs.append(item[0])
	QUAs.append(item[2])
#	goes through the listmap variable above and builds the lists from its items

GPUs	=	list(set(GPUs))
#	for Python to return just the unique elements to a list, you make it a set
#	as sets are different to work with, the unique elements are converted back to being a list and stored to these variables
if 		TYPE	==	"MULTI":
#	checks if the type is multi-GPU, or multi-API
	QUAs	=	[mQUA]
#	instead of letting it pull from list map, the Quality is set to be a specific value provided earlier
elif	TYPE	==	"SINGLE":
#	else if, so if the previous switch was false, it will check if this is true
	QUAs	=	list(set(QUAs))
#		create a list from the set of values in the GPU and Quality lists.

SUBS	=	"NULL"
#	SUBS is for setting the COLUMN value in the R script
#		the default of NULL ensures the default subsetting written into the file.

droppedGame	=	RelPath.rsplit("\\", 3)[1]	\
	.replace(" Performance Analysis", "")	\
	.replace(" Review", "")
#	always the folder above OCAT data is the name for the project
#	the replace functions then remove the article type from the name, if it is a performance analysis or review, getting just the game name
#		for miscellaneous artiles though, the article's name is appropriate to use

if		TYPE	==	"MULTI" and len(GPUs) != 1:
	cGPU	=	"NULL"
elif	TYPE	==	"SINGLE" or len(GPUs) == 1:
	cGPU	=	"\"" + str(GPUs[0]) + "\""
	SUBS	=	"\"Quality\""
#	sets the cGPU, current GPU, variable based on if this is a single or multi-GPU situation
#		quotes are added to the string here, which is why \" is present
#	GPUsread is a list of the read GPUs from the folder names and useful for single-GPU, multi-API situations
#	a single-GPU situation is most likely to call for a specific column to be subset by, and the most likely column is "Quality" then
#		the quotation marks are needed in the R script, so I need to escape them here so they will be placed in script.

scriptFull	=	scriptPath + "OCAT - Search - PA.r"
#	sets the name and path of the reference R script to be loaded in

outputName	=	"Search - PA - " + droppedGame + ".r"
outputFull	=	droppedPath + "@" + outputName
#	sets the name and path of the output file, derived from the reference R script

RPath		=	droppedPath.replace("\\", "/")
#	R uses / instead of \ for paths, so this replaces those characters in the string

if not os.path.exists(outputFull):
#	checks if the output script already exists, so as to not overwrite it
	with open(scriptFull, 'r') as fref, open(outputFull, 'w') as fout:
#		loads the reference script to the variable fref
#		opens an output file to write to, to the variable fout
		for line in fref:
#			loops through the lines in the reference file
			fout.write(line	\
				.replace("!PATH!",		RPath)			\
				.replace("!GAME!",		droppedGame)	\
				.replace("!QUA!",		QUAs[0])
			)
#				writes each line from the reference file to the output file, but runs the replace function to change the appropriate things
#					the \ are necessary to have the line breaks within the command list
		fout.close()
#			closes the output file

scriptFull	=	scriptPath + "OCAT - Combined - Input.r"
#	sets the name and path of the reference R script to be loaded in

outputName	=	"Combined - Input - " + droppedGame + ".r"
outputFull	=	droppedPath + "@" + outputName
#	sets the name and path of the output file, derived from the reference R script

if not os.path.exists(outputFull):
#	checks if the output script already exists, so as to not overwrite it
	with open(scriptFull, 'r') as fref, open(outputFull, 'w') as fout:
#		loads the reference script to the variable fref
#		opens an output file to write to, to the variable fout
		for line in fref:
#			loops through the lines in the reference file
			fout.write(line
				.replace("!PATH!",	RPath)			\
				.replace("!GAME!",	droppedGame)	\
				.replace("!QUA!",	QUAs[0])		\
				.replace("!GPU!",	cGPU)			\
				.replace("!SUBS!",		SUBS
			)
#				writes each line from the reference file to the output file, but runs the replace function to change the appropriate things
		fout.close()
#			closes the output file

if not os.path.exists(droppedPath + "@Combined - Output.r"):
#	checks if the output script already exists, so as to not overwrite it
	shutil.copyfile(scriptPath + "\\OCAT - Combined - Output.r", droppedPath + "@Combined - Output.r")
#		copies the reference file to the appropriate location, and changes the name

# os.system("pause")
#	just used to pause the script instead of closing, for troubleshooting purposes

File: OCAT - Search - PA.r
#	this script is to collect all of the CSVs into one CSV
library(readr)

game	=	"!GAME!"
COMPRESS	=	TRUE
noCOL	=	c("Motherboard","OS","Processor","System RAM","Base Driver Version","Driver Package","GPU #","GPU","GPU Core Clock (MHz)","GPU Memory Clock (MHz)","GPU Memory (MB)")

COLUMN	=	!SUBS!	;	SUBSET	=	"!QUA!"

setwd("!PATH!")
relPath	=	paste0(unlist(strsplit(getwd(), "OCAT Data"))[1], "OCAT Data")

if	(getwd() == relPath & (is.null(COLUMN) | is.null(SUBSET)))	{
	COLUMN	=	"Quality"
	SUBSET	=	"High"
}

txtFIND	=	function(TXT, rel.Path = relPath)	{
	if (file.exists(TXT))	{
		return(readLines(TXT, warn = FALSE))
	}	else	{
		locFILE	=	paste0(rel.Path, "/", TXT)
		if (file.exists(locFILE))	return(readLines(locFILE, warn = FALSE))
	}
	return(NULL)
}

listGPU	=	txtFIND("GPUs.txt")
listQUA	=	txtFIND("Qualities.txt")
listAPI	=	txtFIND("APIs.txt")
listLOC	=	txtFIND("Locations.txt")

csvFIND	=	function(DIRECT = getwd())	{
	LIST		=	list.files(DIRECT, recursive = TRUE, pattern = ".csv")
	LIST		=	LIST[grepl("OCAT", LIST) & grepl("csv", LIST) & !startsWith(LIST, "@")]
	LIST.full	=	paste0(DIRECT, "/", LIST)
	LIST.rel	=	t(data.frame(lapply(LIST.full, strsplit, "OCAT Data/"), row.names = NULL)[2, ])
	colnames(LIST.rel)	=	NULL
	rownames(LIST.rel)	=	NULL

	return(LIST.rel)
}

# LIST.rel	=	csvFIND()


csvCONF	=	function(CSV.list, LOCs = listLOC)	{
	CSV.config	=	t(as.data.frame(sapply(CSV.list, strsplit, "/")))
	colnames(CSV.config)	=	NULL
	rownames(CSV.config)	=	NULL

	CONFIG	=	data.frame(matrix(ncol = 5, nrow = nrow(CSV.config)))
	colnames(CONFIG)	=	c("GPU", "API", "Quality", "Location", "CSV")

	CONFIG$GPU		=	CSV.config[, 1]
	if	(ncol(CSV.config) == 4)	CONFIG$API		=	CSV.config[, 2]
	CONFIG$Quality	=	CSV.config[, ncol(CSV.config)-1]

	appLOC	=	function(DATA, LOCs)	{
		if (is.null(LOCs))	LOCs	=	paste0("Recording ", 1:nrow(DATA))
		rep(LOCs, length.out = nrow(DATA))
	}
	GROUPS	=	list(GPU = CONFIG$GPU, API = CONFIG$API, Quality = CONFIG$Quality)
	if (any(is.na(GROUPS$API)))	GROUPS$API	=	NULL
	CONFIG$Location	=	unlist(by(CONFIG, GROUPS, appLOC, LOCs))
#	appLOC and by together apply the Location names, and generate them if necessary, to match the GPU-API-Quality groups
	CONFIG$CSV		=	CSV.config[, ncol(CSV.config)]

	return(CONFIG)
}

# CSV.config	=	csvCONF(LIST.rel)
CSV.configFull	=	csvCONF(csvFIND())

csvFILT	=	function(CSV.list, COL, SUB)	{
	if (!is.null(COL)	&	!is.null(SUB))	return(CSV.list[CSV.list[, COL] == SUB, ])
	return(CSV.list)
}

if (SUBSET == "All")	CSV.config	=	CSV.configFull
if (SUBSET != "All")	CSV.config	=	csvFILT(CSV.configFull, COLUMN, SUBSET)


read_OCAT	=	function(INFO, GPU = NULL, API = NULL, QUA = NULL, LOC = NULL)	{
	if (is.data.frame(INFO))	{
		GPU		=	INFO$GPU
		API		=	INFO$API
		QUA		=	INFO$Quality
		LOC		=	INFO$Location
		FILE	=	INFO$CSV
	}	else	{FILE	=	INFO}

	filePATH		=	paste(relPath, GPU, QUA, FILE, sep = "/")
	if	(!any(is.na(API), is.null(API)))	filePATH		=	paste(relPath, GPU, API, QUA, FILE, sep = "/")

	out				=	read_csv(filePATH, guess_max = 10, lazy = TRUE, col_select=!all_of(noCOL), show_col_type = FALSE)

	out$GPU			=	GPU
	out$Quality		=	QUA
	out$Location	=	LOC
	out$API			=	""
	if	(!any(is.na(API), is.null(API)))	out$API	=	API

	return(out)
}


csvOCAT	=	function(CSVs)	{
	OCATcomb	=	data.frame(matrix(ncol = 24, nrow = 0))
	for (ROW in 1:nrow(CSVs))	{
		OCATcomb	=	rbind(OCATcomb, read_OCAT(CSVs[ROW, ]))
	}

	return(OCATcomb)
}

if (!is.null(listGPU))	CSV.config$GPU		=	ordered(CSV.config$GPU,			listGPU)
if (!is.null(listAPI))	CSV.config$API		=	ordered(CSV.config$API,			listAPI)
if (!is.null(listQUA))	CSV.config$Quality	=	ordered(CSV.config$Quality,		listQUA)
if (!is.null(listLOC))	CSV.config$Location	=	ordered(CSV.config$Location,	listLOC)

CSV.config	=	CSV.config[order(CSV.config$GPU, CSV.config$API, CSV.config$Quality, CSV.config$Location), ]
OCATcomb	=	csvOCAT(CSV.config)

PREFIX	=	"@Combined"
if (SUBSET == "All")	PREFIX	=	game

if	(COMPRESS)	{
	write_csv(OCATcomb, paste0(PREFIX, " - ", SUBSET, ".csv.bz2"))
}	else	{
	write_csv(OCATcomb, paste0(PREFIX, " - ", SUBSET, ".csv"))
}